contracts/MetawoodAuction.sol

```diff
@@ -0,0 +1,330 @@
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract MetawoodAuction is ERC1155Holder, Ownable, ReentrancyGuard {
+    using SafeMath for uint256;
+    using Counters for Counters.Counter;
+
+    Counters.Counter private _auctionCounter;
+    IMetawoodNFT public metawoodNFT;
+
+    uint256 public bidIncreaseThreshold = 200;
+
+    constructor(address _nftContract) {
+        metawoodNFT = IMetawoodNFT(_nftContract);
+    }
+
+    enum AuctionState {
+        CLOSED,
+        ACTIVE
+    }
+
+    //single struct for handling bid, tbd?
+    struct Auction {
+        uint256 id;
+        IMetawoodNFT nftContract;
+        uint256 nftId; // NFT Id
+        address payable creator; // Creator of the Auction
+        uint256 bidIncreaseThreshold;
+        uint256 auctionEnd;
+        uint256 minimumBid;
+        uint256 highestBid;
+        address payable highestBidder;
+        uint256[] bids;
+        address[] bidders;
+        // address payable[] bidders;
+        AuctionState status;
+    }
+
+    mapping(uint256 => Auction) private _auctions;
+
+    event AuctionCreated(
+        uint256 indexed auctionId,
+        address nftContract,
+        uint256 indexed nftId,
+        address indexed creator,
+        uint256 bidIncreaseThreshold,
+        uint256 auctionEnd,
+        uint256 minimumBid
+    );
+
+    event BidReceived(uint256 indexed auctionId, uint256 bidAmount, address indexed bidder);
+    event AuctionSettled(uint256 indexed auctionId);
+    event AuctionClosed(uint256 indexed auctionId);
+    event AuctionEndUpdated(uint256 indexed auctionId, uint256 updatedTime);
+
+    modifier ensureNonZeroAddress(address addressToCheck) {
+        require(addressToCheck != address(0), "Zero address!");
+        _;
+    }
+
+    modifier ensureNFTOwner(uint256 _tokenId) {
+        require(metawoodNFT.exists(_tokenId), "Non existing token");
+        require(metawoodNFT.balanceOf(msg.sender, _tokenId) > 0, "Token Not Owned!");
+        _;
+    }
+
+    modifier ensureValidAuction(uint256 _auctionId) {
+        Auction memory auction = _auctions[_auctionId];
+        require(auction.creator != address(0), "Invalid Auction");
+        require(msg.sender != address(0), "Invalid caller address");
+        _;
+    }
+
+    modifier ensureActiveAuction(uint256 _auctionId) {
+        Auction memory auction = _auctions[_auctionId];
+        require(auction.status == AuctionState.ACTIVE, "Auction not active");
+        _;
+    }
+
+    modifier ensureAuctionCreator(uint256 _auctionId) {
+        Auction memory auction = _auctions[_auctionId];
+        require(auction.creator == msg.sender, "Not the auction creator");
+        _;
+    }
+
+    // Transactions
```

Comment:
Include the constructor below the modifers and events.
And after the constructor, add all the read and write functions.
Btw nice work on the events and modifers

Reply:
Thanks, will do that :+1: